052-record-reconciled-versions.md

```diff
@@ -0,0 +1,160 @@
+e.g. then 
+```
+strimziReconciled: null
+strimziManagedBy: null
+```
+would in a real scenario simply be:
+```
+status: {}
+```
+
+### Fresh Install
+
+- On a fresh install, a user creates a Kafka CR:
+  ```
+  strimziReconciled: null
+  strimziManagedBy: null
+  ```
+- The Kafka Reconciler watcher picks up the CR does an initial status update:
+  ```
+  strimziReconciled: null
+  strimziManagedBy: 0.37.0
+  ```
+  and creates `StrimziPodSet` CRs
+- StrimziPodSet reconciler reconciles the `SPS`s
+- Kafka Reconciler picks up that `SPS` is now Ready and finished installing.
+  Kafka Reconciler continues to install other components such as entity operator.
+  Once all components are deployed and updated, update `Kafka` CR with
+  ```
+  strimziReconciled: 0.37.0
+  strimziManagedBy: 0.37.0
+  ```
+- User can now verify `Kafka.status.versions.strimziReconciled` to check if it has reconciled to correct/expected level.
+- if there is an error at any point during reconcile `Kafka.status.versions.strimziReconciled` will not be updated.
+
+
+### Upgrade (from no mechansim)
+Same as fresh install, upgrade from 0.37.0 to 0.38.0
+
+- User has pre-existing Kafka CR:
+  ```
+  strimziReconciled: null
+  strimziManagedBy: null
+  ```
+- The Kafka Reconciler watcher picks up the CR does an initial status update:
+  ```
+  strimziReconciled: null
+  strimziManagedBy: 0.38.0
+  ```
+  and patches `StrimziPodSet` CRs
+- StrimziPodSet reconciler reconciles the `SPS`s with help of `KafkaRoller` and `ZookeeperRoller`
+- Kafka Reconciler picks up that `SPS` is now Ready and finished upgrading.
+  Kafka Reconciler continues to install other components such as entity operator.
+  Once all components are upgraded, update `Kafka` CR with
+  ```
+  strimziReconciled: 0.38.0
+  strimziManagedBy: 0.38.0
+  ```
+- User can now verify `Kafka.status.versions.strimziReconciled` to check if it has reconciled to correct/expected level.
+- if there is an error at any point during reconcile `Kafka.status.versions.strimziReconciled` will not be updated.
+
+
+### Upgrade (with new mechansim)
+This defines an example where it is imagined that `0.37.0` had this mechanism already implemented, and is now upgrading to a new version 0.38.0
+
+- User has pre-existing Kafka CR:
+  ```
+  strimziReconciled: 0.37.0
+  strimziManagedBy: 0.37.0
+  ```
+- The Kafka Reconciler watcher picks up the CR does an initial status update:
+  ```
+  strimziReconciled: 0.37.0
+  strimziManagedBy: 0.38.0
+  ```
+  and patches `StrimziPodSet` CRs
+- StrimziPodSet reconciler reconciles the `SPS`s with help of `KafkaRoller` and `ZookeeperRoller`
+- Kafka Reconciler picks up that `SPS` is now Ready and finished upgrading.
+  Kafka Reconciler continues to install other components such as entity operator.
+  Once all components are upgraded, update `Kafka` CR with
+  ```
+  strimziReconciled: 0.38.0
+  strimziManagedBy: 0.38.0
```

Comment:
Did you abandon the idea of annotating the StrimziPodSets and Deployments? It is a bit confusing to me as you mention the rollers here or the SPS / Deployments int he intro. But there are no more details about them.

Reply:
I did abandon the idea somewhat after it was made clear to me, after our conversation that the KafkaAssembly had all the info it needed to make the determination whether a full roll/reconcile was finished (if this is not the case though, please correct me on this).

Technically we could continue to propagate the annotation through the pods, but considering the main aim of the proposal is to have info in the status to reflect whether the upgrade was complete, I think this metadata should only be propagated to the deployments and SPS if it is required as part of the mechanism of determining whether a full reconcile has finished.

The purpose is to make an upgrades Readiness + finished state more clear to a user in just the source of truth, the `Kafka` CR rather than relying on checking the pods directly.